### PR TITLE
[Bugfix] Fix CFG-parallel broadcast dtype mismatch in GLM-Image

### PR DESCRIPTION
**Pull request by dsocek**
## Purpose

Fix CFG-parallel deadlock/crash in GLM-Image. In the CFG-parallel path, rank 0 latents become `float32` after `scheduler.step()` (inherited from `noise_pred`), while rank 1 latents remain at `transformer_dtype` (e.g. `bfloat16`). The subsequent `cfg_group.broadcast(latents, src=0)` then violates the collective’s same-dtype requirement.

Fix: cast latents back to `transformer_dtype` on rank 0 before the broadcast.

## Test Plan

Run GLM-Image with CFG parallelism enabled on 2 GPUs and submit a text-to-image request.

## Test Result

Reproduced on 2x Intel Arc B60 (XPU, oneCCL): without the fix, the very first diffusion step hangs in `torch.distributed.broadcast`. With the fix, all 28 diffusion steps complete at as expected.

### CODE REVIEW

**Comment by dsocek** [comment]
@princepride Friendly ping on this PR when you have a chance for a quick review. This is a small bug fix for a CFG-parallelism hang in GLM-Image. Thanks!

**Comment by princepride** [comment]
@JaredforReal ptal

**Comment by xuechendi** [comment]
Thanks for the PR, @dsocek , this is an known issue with oneCCL, Fix looks good to me. @hsliuustc0106

**Comment by dsocek** [comment]
@JaredforReal @hsliuustc0106 Checking back on this PR. Since @xuechendi confirmed the fix looks good and matches a known oneCCL issue, I was wondering if there's anything else needed to move it forward. Thanks!

**Review by RuixiangMa — COMMENTED** [review]
LGTM

**Comment by herotai214** [comment]
LGTM;
I've merged this PR with latest commit (`77d395`, Jun 17)

Briefly tested on H800, CFG-parallel for GLM-Image runnable with below results:
With `cfg_parallel_size: 2`: e2e_wall_time_ms = 37,670.123

```
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] [Overall Summary]
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] +-----------------------------+------------+
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | Field                       |      Value |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] +-----------------------------+------------+
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_requests                |          1 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_wall_time_ms            | 37,670.123 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_total_tokens            |      2,064 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_avg_time_per_request_ms | 37,670.123 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_avg_tokens_per_s        |     54.791 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | input_preprocess_time_ms    |     98.754 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_stage_0_wall_time_ms    | 18,850.769 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] | e2e_stage_1_wall_time_ms    | 18,584.042 |
[0;36m(APIServer pid=556502)[0;0m INFO 06-18 07:54:36 [stats.py:758] +-----------------------------+------------+
```

With `cfg_parallel_size: 1`: e2e_wall_time_ms = 53,173.269 
```
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] [Overall Summary]
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] +-----------------------------+------------+
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | Field                       |      Value |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] +-----------------------------+------------+
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_requests                |          1 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_wall_time_ms            | 53,173.269 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_total_tokens            |      2,064 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_avg_time_per_request_ms | 53,173.269 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_avg_tokens_per_s        |     38.816 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | input_preprocess_time_ms    |    106.951 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_stage_0_wall_time_ms    | 18,562.336 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] | e2e_stage_1_wall_time_ms    | 34,402.355 |
[0;36m(APIServer pid=559932)[0;0m INFO 06-18 08:14:22 [stats.py:758] +-----------------------------+------------+
```


Script:
```
# Set $DEPLOY_CONFIG to the GLM-Image yaml file to (not) activate cfg parallel
vllm serve "$MODEL" --omni \
    --deploy-config "$DEPLOY_CONFIG" \
    --port "$PORT"

# Send request
python examples/online_serving/image_to_image/openai_chat_client.py \
    --server http://localhost:$PORT\
    --prompt "Convert this image to watercolor style" \
    --input $INPUT \
    --guidance 1.5 \ # guidance > 1.0 to activate cfg
    --output $OUTPUT
```

Note that there seems no explict logs to indicate cfg related behaviour for GLM-Image, but logs like `Stage 1 logical-to-physical device mapping: 1->5, 2->6` and the accelerated e2e results show the cfg parallel should be working here.

**Comment by hsliuustc0106** [comment]
@lyj-jjj PTAL

**Comment by dsocek** [comment]
@lyj-jjj @JaredforReal Checking back on this PR —without this fix, CFG-parallel GLM-Image hangs on the very first diffusion step (deadlock in `torch.distributed.broadcast`), making the feature unusable.

The fix is minimal (2 lines) and has been validated across multiple hardware configs:

- 2×Intel Arc B60 (XPU/oneCCL) — fix unblocks the hang
- 2×H200 (CUDA) — ~30% e2e speedup; DiT component is ~2x faster with cfg_parallel_size=2
- 2×B70 (XPU) — working as expected with the fix, DiT component is ~2x faster with cfg_parallel_size=2

Multiple reviewers have already LGTM'd (@RuixiangMa, @xuechendi, @herotai214). Any blockers to merge?

**Comment by lyj-jjj** [comment]
@hsliuustc0106 This fix is reasonable and can be merged.

**Review by wtomin — APPROVED** [review]
This fix looks good to me and the experiment results confirm that the CFG parallel feature works as expected.
